subsys/net/lib/openthread/platform/uart.c

```diff
@@ -0,0 +1,220 @@
+	RING_BUF_DECLARE(_name##_tx_ringbuf, _ringbuf_size); \
+	RING_BUF_DECLARE(_name##_rx_ringbuf, _ringbuf_size); \
+	static struct openthread_uart _name = { \
+		.tx_ringbuf = &_name##_tx_ringbuf, \
+		.rx_ringbuf = &_name##_rx_ringbuf, \
+	};
+
+OT_UART_DEFINE(ot_uart, CONFIG_OPENTHREAD_NCP_UART_RING_BUFFER_SIZE)
+
+#define RX_FIFO_SIZE 128
+
+static void uart_rx_handle(/*const struct openthread_uart *ot_uart*/)
+{
+	u8_t *data;
+	u32_t len;
+	u32_t rd_len;
+	bool new_data = false;
+
+	do {
+		len = ring_buf_put_claim(
+			ot_uart.rx_ringbuf, &data,
+			ot_uart.rx_ringbuf->size);
+		if (len > 0) {
+			rd_len = uart_fifo_read(
+				ot_uart.dev, data, len);
+
+			if (rd_len > 0) {
+				new_data = true;
+			}
+
+			int err = ring_buf_put_finish(
+				ot_uart.rx_ringbuf, rd_len);
+			(void)err;
+			__ASSERT_NO_MSG(err == 0);
+		} else {
+			u8_t dummy;
+
+			/* No space in the ring buffer - consume byte. */
+			LOG_WRN("RX ring buffer full.");
+
+			rd_len = uart_fifo_read(
+				ot_uart.dev, &dummy, 1);
+		}
+	} while (rd_len && (rd_len == len));
+
+	if (new_data) {
+		otSysEventSignalPending();
+	}
+}
+
+static void uart_tx_handle(/*struct openthread_uart *ot_uart*/)
+{
+	u32_t len;
+
+	const u8_t *data;
+
+	len = ring_buf_get_claim(ot_uart.tx_ringbuf, (u8_t **)&data,
+				 ot_uart.tx_ringbuf->size);
+	if (len) {
+		int err;
+
+		len = uart_fifo_fill(ot_uart.dev, data, len);
+		err = ring_buf_get_finish(ot_uart.tx_ringbuf, len);
+		(void)err;
+		__ASSERT_NO_MSG(err == 0);
+	} else {
+		uart_irq_tx_disable(ot_uart.dev);
+		ot_uart.tx_busy = 0;
+		atomic_set(&(ot_uart.tx_finished), 1);
+		otSysEventSignalPending();
+	}
+}
+
+static void uart_callback(void *user_data)
+{
+	while (uart_irq_update(ot_uart.dev) &&
+	       uart_irq_is_pending(ot_uart.dev)) {
+
+		if (uart_irq_rx_ready(ot_uart.dev)) {
+			uart_rx_handle();
+		}
+
+		if (uart_irq_tx_ready(ot_uart.dev)) {
+			uart_tx_handle();
+		}
+	}
+};
+
+void platformUartProcess(otInstance *aInstance)
+{
+	u32_t len = 0;
+	const u8_t *data;
+
+	/* Process UART RX */
+	while (
+				(
+				len = ring_buf_get_claim(
+						ot_uart.rx_ringbuf,
+						(u8_t **)&data,
+						ot_uart.rx_ringbuf->size)
+				) > 0
+	) {
+		otPlatUartReceived(data, len);
+		int err = ring_buf_get_finish(
+				ot_uart.rx_ringbuf,
+				len);
+		(void)err;
+		__ASSERT_NO_MSG(err == 0);
+	}
+
+	/* Process UART TX */
+	if (ot_uart.tx_finished) {
+		LOG_DBG("UART TX done");
+		otPlatUartSendDone();
+		ot_uart.tx_finished = 0;
+	}
+};
+
+otError otPlatUartEnable(void)
+{
+	ot_uart.dev = device_get_binding(
+		CONFIG_OPENTHREAD_NCP_SPINEL_ON_UART_DEV_NAME);
+
+	if ((&ot_uart)->dev == NULL) {
+		LOG_ERR("UART device not found");
+		return OT_ERROR_FAILED;
+	}
+
+#ifdef CONFIG_OPENTHREAD_NCP_SPINEL_ON_UART_ACM
+	int ret = usb_enable(NULL);
+
+	if (ret != 0) {
+		LOG_ERR("Failed to enable USB");
+		return OT_ERROR_FAILED;
+	}
+
+	/* Wait 1 sec for the host to do all settings */
+	LOG_INF("Wait for host to settle");
+	k_busy_wait(1000000);
```

Comment:
Is the `k_busy_wait` really needed here, couldn't we let other tasks work in the meantime (as it would happen with `k_sleep`)?

Reply:
This is the same as in https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/subsys/usb/cdc_acm/src/main.c#L119.

Reply:
Even if it is like that in other sample, it does not mean it is correct thing to do. I would like to understand why we should have `k_busy_wait()` here instead of `k_sleep()`. Have you tried to use the k_sleep(), does it cause issues?

Reply:
Tested just now on hardware with `k_busy_wait()` wpantund starts up just fine. With `k_sleep()` wpantund goes into 
```
wpan0 => [
        "NCP:State" => "uninitialized:fault"
```

Not an expert in the USB stack. So, I would keep it like this and someone knowledgeable from the USB team to take a look into this question here and in the CDC-ACM sample.

Reply:
Ok, that is good info. Could you add a comment to the code about your findings so we do not need wonder later why the busy wait is being used there.

Reply:
Done.

Reply:
If k_sleep() does not work then there is probably a bug or issue, anyway I suggest to change it to `k_busy_wait(USEC_PER_SEC * 1);`

Reply:
I did not notice the different time-base between `k_busy_wait(us)` and `k_sleep(ms)`. Once moving to a millisecond sleep, it also works with `k_sleep()`. Fixed and rebased.